lib/Lower/Bridge.cpp

```diff
@@ -574,41 +380,328 @@ class FirConverter : public Fortran::lower::AbstractConverter {
+        : loopVariableSym{sym}, lowerExpr{lowerExpr}, upperExpr{upperExpr},
+          stepExpr{stepExpr}, loopVariableType{type} {}
+
+    bool isStructured() const { return headerBlock == nullptr; }
+
+    // Data members for both structured and unstructured loops.
+    Fortran::semantics::Symbol *loopVariableSym;
+    const Fortran::parser::ScalarExpr &lowerExpr;
+    const Fortran::parser::ScalarExpr &upperExpr;
+    const std::optional<Fortran::parser::ScalarExpr> &stepExpr;
+    mlir::Type loopVariableType;
+    mlir::Value loopVariable{mlir::Value{}};
+    mlir::Value stepValue{mlir::Value{}}; // possible uses in multiple blocks
+
+    // Data members for structured loops.
+    fir::LoopOp doLoop{};
+    mlir::OpBuilder::InsertPoint insertionPoint{};
+
+    // Data members for unstructured loops.
+    mlir::Value tripVariable{mlir::Value{}};
+    mlir::Block *headerBlock{nullptr};    // loop entry and test block
+    mlir::Block *bodyBlock{nullptr};      // first loop body block
+    mlir::Block *successorBlock{nullptr}; // loop exit target block
+  };
+
+  /// Generate FIR for a DO construct.  There are six variants:
+  ///  - unstructured infinite and while loops
+  ///  - structured and unstructured increment loops
+  ///  - structured and unstructured concurrent loops
+  void genFIR(Fortran::lower::pft::Evaluation &eval,
+              const Fortran::parser::DoConstruct &) {
+    bool unstructuredContext = eval.lowerAsUnstructured();
+    Fortran::lower::pft::Evaluation &doStmtEval = eval.evaluationList->front();
+    auto *doStmt = doStmtEval.getIf<Fortran::parser::NonLabelDoStmt>();
+    assert(doStmt && "missing DO statement");
+    const auto &loopControl =
+        std::get<std::optional<Fortran::parser::LoopControl>>(doStmt->t);
+    llvm::SmallVector<IncrementLoopInfo, 1> incrementLoopInfo;
+    const Fortran::parser::ScalarLogicalExpr *whileCondition = nullptr;
+    bool infiniteLoop = !loopControl.has_value();
+    if (infiniteLoop) {
+      assert(unstructuredContext && "infinite loop must be unstructured");
+      startBlock(doStmtEval.localBlocks[0]); // header block
+    } else if ((whileCondition =
+                    std::get_if<Fortran::parser::ScalarLogicalExpr>(
+                        &loopControl->u))) {
+      assert(unstructuredContext && "while loop must be unstructured");
+      startBlock(doStmtEval.localBlocks[0]); // header block
+      genFIRConditionalBranch(*whileCondition, doStmtEval.lexicalSuccessor,
+                              doStmtEval.parentConstruct->constructExit);
+    } else if (const auto *bounds =
+                   std::get_if<Fortran::parser::LoopControl::Bounds>(
+                       &loopControl->u)) {
+      // "Normal" increment loop.
+      incrementLoopInfo.emplace_back(IncrementLoopInfo(
+          bounds->name.thing.symbol, bounds->lower, bounds->upper, bounds->step,
+          genType(*bounds->name.thing.symbol), builder->getInsertionBlock()));
```

Comment:
You shouldn't need to use a constructor here. `emplace_back` ought to forward the argument list to the constructor to skip copying a temporary into the container.

Reply:
Thanks